pkg/approver/aws/csr_test.go

```diff
@@ -0,0 +1,636 @@
+				certificates.UsageDigitalSignature,
+				certificates.UsageClientAuth,
+			},
+		}
+		c.cbm(&b)
+		t.Run(fmt.Sprintf("test #%d", idx), func(t *testing.T) {
+			fclient := &fake.Clientset{}
+			fclient.AddReactor("get", "nodes", c.react)
+			ar := &Approver{
+				kubeClient: fclient,
+			}
+			csr := makeTestCsr(b)
+			x509cr, err := csrutil.ParseCSR(csr)
+			if err != nil {
+				t.Errorf("unexpected err: %v", err)
+			}
+			r := ar.isValidNode(mapNodesToInstanceIDs(nodetoid))
+			err = r(csr, x509cr)
+			if err != nil && !c.err {
+				t.Errorf("expected err: %v got:%v", c.err, err)
+			}
+			c.verify(t, fclient.Actions())
+		})
+	}
+}
+
+func TestIsValidASG(t *testing.T) {
+	nodetoasg := map[string]string{
+		"valid-node":   "asg-1",
+		"another-node": "asg-2",
+	}
+	wasgs := []string{"asg-1"}
+	cases := []struct {
+		cbm func(b *csrBuilder)
+		err bool
+	}{{
+		cbm: func(b *csrBuilder) {
+			b.cn = "invalid"
+		},
+		err: true,
+	}, {
+		cbm: func(b *csrBuilder) {
+			b.cn = "system:node:random-node"
+		},
+		err: true,
+	}, {
+		cbm: func(b *csrBuilder) {
+			b.cn = "system:node:another-node"
+		},
+		err: true,
+	}, {
+		cbm: func(b *csrBuilder) {},
+		err: false,
+	}}
+
+	for idx, c := range cases {
+		b := csrBuilder{
+			cn:              "system:node:valid-node",
+			orgs:            []string{"system:nodes"},
+			requestor:       "system:node:valid-node",
+			requestorGroups: []string{"system:nodes"},
+			usages: []certificates.KeyUsage{
+				certificates.UsageKeyEncipherment,
+				certificates.UsageDigitalSignature,
+				certificates.UsageClientAuth,
+			},
+		}
+		c.cbm(&b)
+		t.Run(fmt.Sprintf("test #%d", idx), func(t *testing.T) {
+			ar := &Approver{
+				allowedASGs: sets.NewString(wasgs...),
+			}
+			csr := makeTestCsr(b)
+			x509cr, err := csrutil.ParseCSR(csr)
+			if err != nil {
+				t.Errorf("unexpected err: %v", err)
+			}
+			r := ar.isValidASG(mapNodesToASGs(nodetoasg))
+			err = r(csr, x509cr)
+			if err != nil && !c.err {
+				t.Errorf("expected err: %v got:%v", c.err, err)
+			}
+		})
+	}
+}
+
+func mapNodesToInstanceIDs(m map[string]string) instanceIDFunc {
+	return func(nodeName string) (string, error) {
+		id, ok := m[nodeName]
+		if !ok {
+			return "", fmt.Errorf("no instance id found for %s", nodeName)
+		}
+		return id, nil
+	}
+}
+
+func mapNodesToASGs(m map[string]string) autoScalingGroupIDFunc {
+	return func(nodeName string) (string, error) {
+		id, ok := m[nodeName]
+		if !ok {
+			return "", fmt.Errorf("no asg found for %s", nodeName)
+		}
+		return id, nil
+	}
+}
+
+type csrBuilder struct {
+	cn              string
+	orgs            []string
+	requestor       string
+	requestorGroups []string
+	usages          []certificates.KeyUsage
+	dns             []string
+	emails          []string
+	ips             []net.IP
+}
+
+func makeTestCsr(b csrBuilder) *certificates.CertificateSigningRequest {
+	random := rand.Reader
+	pk, err := ecdsa.GenerateKey(elliptic.P224(), random)
+	if err != nil {
+		panic(err)
```

Comment:
nit: instead of panicing pass a testing.T to makeTestCSR and call t.Fatal

```
func makeTestCSR(t *testing.T, b csrBuilder) ... {
    // ...

    if err != nil {
        t.Fatal(err)
    }
}
```

Reply:
done